### PR TITLE
fix: escape filename pipes in diagnostic.lua

### PR DESCRIPTION
**Pull request by mtoohey31**
Hello! I noticed, while working on a file with a pipe (`|`) in it's name and an LSP running, that there were errors being throw in the lua file I've modified here. It looks like the formatted command didn't have pipes escaped correctly, so when it was evaluated, the autocommand's argument would only be the filename up to the point of the pipe, then everything after that point would be treated as a second command to be evaluated.

The change I made does some simple escaping, and works as expected after testing, though I'm not sure if it handles every edge case; I'm not a vimscript expert. I'm also not familiar with the codebase, so lmk if there's already a function somewhere for escaping this kind of value, and I'd be happy to update this solution to use that!

### CODE REVIEW

**Comment by zeertzjq** [comment]
Maybe using `vim.fn.fnameescape()` is better.

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -658,7 +658,7 @@ function M.set(namespace, bufnr, diagnostics, opts)
 
   vim.api.nvim_buf_call(bufnr, function()
     vim.api.nvim_command(
-      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.api.nvim_buf_get_name(bufnr))
+      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.api.nvim_buf_get_name(bufnr):gsub("|", "\\|"))
```
Comment (zeertzjq) [inline runtime/lua/vim/diagnostic.lua:661]: Maybe using `vim.fn.fnameescape()` is better.
```suggestion
      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.fn.fnameescape(vim.api.nvim_buf_get_name(bufnr)))
```

**Comment by zeertzjq** [comment]
There is also a similar call in line 1336. You may want to change that as well.

**Comment by gpanders** [comment]
A test would be good as well to demonstrate the failure case and show that this change fixes it.

File: runtime/lua/vim/diagnostic.lua (hunk 1 above)
Comment (mtoohey31) [reply runtime/lua/vim/diagnostic.lua]: Thank you, fixed in c52aa7c.

**Comment by mtoohey31** [comment]
> A test would be good as well to demonstrate the failure case and show that this change fixes it.

Sounds good, I'm on it!

**Comment by gpanders** [comment]
> > A test would be good as well to demonstrate the failure case and show that this change fixes it.
> 
> Sounds good, I'm on it!

`diagnostic_spec.lua` is the place to add it. There are plenty of existing tests for you to ~~copy~~ draw inspiration from.

**Comment by mtoohey31** [comment]
Alright, I believe 542293a covers things, testing with and without the `fnameescape` change, both tests fail beforehand, and succeed after.

File: test/functional/lua/diagnostic_spec.lua [hunk 2]
```diff
@@ -1918,5 +1918,31 @@ describe('vim.diagnostic', function()
         return {show_called, hide_called}
       ]])
     end)
+
+    it('triggers the autocommand when diagnostics are set', function()
+      eq(1, exec_lua [[
+        vim.g.diagnostic_autocmd_triggered = 0
+        vim.cmd('autocmd DiagnosticChanged * let g:diagnostic_autocmd_triggered = 1')
+        vim.api.nvim_buf_set_name(diagnostic_bufnr, "test | test")
+        vim.wait(100, function ()
```
Comment (gpanders) [inline test/functional/lua/diagnostic_spec.lua:1927]: What's the rationale for `vim.wait()`? Seems to work fine without it.

Comment (mtoohey31) [reply test/functional/lua/diagnostic_spec.lua]: You're right! My bad, it seems it's unecessary. Fixed in 7ad910d.

**Comment by gpanders** [comment]
Thanks for your contribution!

**Comment by mtoohey31** [comment]
No problem, thanks for helping me through things!

**Comment by zeertzjq** [comment]
@gpanders Should this be backported?

**Comment by gpanders** [comment]
Yes, thanks

**Pull request merged**